File: antia_numerical_methods/c/xmp/invit.c
/*	Real eigenvalue and eigenvector of a general real matrix using inverse iteration */

#include <stdio.h>
#include <math.h>

int gauelm(int n, int num, double * a, double * x, double * det,
       int inc[], int lj, int * iflg);
int invit(double *a, int m, int ia, double *p, double u[], int iflg,
	double *ei, double *erc, double reps, int *nit);
int invit_l(double *a, int m, int ia, double *p, double u[], int iflg,
	double *ei, double *erc, double reps, int *nit);

main()
{
	int i,i1,j,n,m, id, iflg, ier,np,nmax;
	double hh, x[100], p,ei,erc,reps,p0;
/*	double a[4][4]={4,-5,0,3, 0,4,-3,-5, 5,-3,4,0, 3,0,5,4}; */
	double a[4][4]={-2,2,2,2, -3,3,2,2, -2,0,4,2, -1,0,0,5};

	id=4; m=n=4; np=0; reps=1.e-14;
	printf(" The matrix is :\n");
	for(i=0; i<n; ++i) {
		for(j=0; j<n; ++j) printf(" %e ",a[i][j]);
		printf("\n");
	}

	for(i1=0; i1<99; ++i1) {
		printf("type iflg=0/1/2,  p=initial shift    (quits when iflg<0)\n");
		scanf(" %d %le",  &iflg, &p);
		if(iflg<0) return 0;

		p0=p;
/*	Initial guess for eigenvector */
		for(i=0; i<m; ++i) x[i]=1.0;
		i=invit(&a[0][0],m,id,&p,x,iflg,&ei,&erc,reps,&np);
		printf(" ier = %d    m =  %d    initial shift = %e \n",i,m,p0);
		printf(" final shift = %e    eigenvalue = %e    Rayleigh Quotient = %e\n",p,ei,erc);
		printf(" eigenvector : ");
		for(i=0; i<m; ++i) printf(" %e ",x[i]);
		printf(" \n");

/*	Calculate the left eigenvector for the same matrix */
		p=p0;
		for(i=0; i<m; ++i) x[i]=1.0;
		i=invit_l(&a[0][0],m,id,&p,x,iflg,&ei,&erc,reps,&np);
		printf(" ier = %d   final shift = %e    eigenvalue = %e    Rayleigh Quotient = %e\n",i,p,ei,erc);
		printf(" left eigenvector : ");
		for(i=0; i<m; ++i) printf(" %e ",x[i]);
		printf(" \n");
	}
	return;
}



/*	Solution of a system of linear equations using Gaussian elimination
	with partial pivoting

	N : (input) Number of equations to be solved
	NUM : (input) Number of different sets (each with N equations) of
	         equations to be solved
	A : (input/output) The matrix of coefficient of size LJ*N
	        A[i][j] is the coefficient of x_j in ith equation
	     	at output it will contain the triangular decomposition
	X : (input/output) The matrix containing right hand sides (size LJ*NUM)
	        X[j][i] is the ith element of jth right hand side
	     	at output it will contain the solutions
	DET : (output) The determinant of the matrix
	INC : (output) Integer array of length N containing information about
		interchanges performed during elimination
	LJ : (input) Second dimension of arrays A and X in calling function
	IFLG : (input) Integer parameter to specify the type of computation required
		If IFLG<=0, both elimination and solution are
			done and IFLG is set to 2
		If IFLG=1, only elimination is done and IFLG is set to 2
		If IFLG>=2 only solution is calculated, the triangular
		    decomposition should have been calculated earlier
		
	Error status is returned by the value of the function GAUELM.
		0 value implies successful execution
		101 implies (N<=0 or N>LJ) 
		121 implies some pivot turned out to be zero and hence
			matrix must be nearly singular

	Required functions : None
*/

#include <math.h>

int gauelm(int n, int num, double * a, double * x, double * det,
       int inc[], int lj, int * iflg)

{
int i,j,k,km,l;
double r1,t1;

	if(n<=0 || n>lj) return 101;
 
	if((*iflg)<2) {
/*	Perform elimination  */
		*det=1.0;
		for(k=0; k<n-1; ++k) {
/*	Find the maximum element in the Kth column  */
			r1=0.0; km=k;
			for(l=k; l<n; ++l)
				if(fabs(a[l*lj+k])>r1) {r1=fabs(a[l*lj+k]); km=l;}

			inc[k]=km;
			if(km != k) {
/*	Interchange the rows if needed  */
				for(l=k; l<n; ++l) 
				{t1=a[k*lj+l]; a[k*lj+l]=a[km*lj+l]; a[km*lj+l]=t1;}
				*det=-(*det);
			}

			*det=(*det)*a[k*lj+k];
			if(a[k*lj+k]==0) return 121;

			for(l=k+1; l<n; ++l) {
				a[l*lj+k]=a[l*lj+k]/a[k*lj+k];
				for(i=k+1; i<n; ++i) a[l*lj+i]=a[l*lj+i]-a[l*lj+k]*a[k*lj+i];
			}
		}
		*det=(*det)*a[(n-1)*lj+n-1];
		inc[n-1]=n-1;
		if(a[(n-1)*lj+n-1]==0) return 121;

		if((*iflg)==1) {*iflg=2; return 0;}
		*iflg=2;
	}

/*	Solution for the num different right-hand sides  */
	for(j=0; j<num; ++j) {
/*	forward-substitution  */
		for(k=0; k<n-1; ++k) {
			if(k != inc[k])
			{t1=x[j*lj+k]; x[j*lj+k]=x[j*lj+inc[k]]; x[j*lj+inc[k]]=t1;}
			for(l=k+1; l<n; ++l) x[j*lj+l]=x[j*lj+l]-a[l*lj+k]*x[j*lj+k];
		}

/*	back-substitution  */

		x[j*lj+n-1]=x[j*lj+n-1]/a[(n-1)*lj+n-1];
		for(k=n-2; k>=0; --k) {
			for(l=n-1; l>=k+1; --l) x[j*lj+k]=x[j*lj+k]-x[j*lj+l]*a[k*lj+l];
			x[j*lj+k]=x[j*lj+k]/a[k*lj+k];
		}
	}
	return 0;
}



/*	Real eigenvalues and eigenvector of a real matrix using inverse iteration

	A : (input) Array of length IA*M containing the matrix elements
	M : (input) Order of the matrix
	IA : (input) The second dimension of A as declared in the calling function, IA>=M
	P : (input/output) Initial value of the shift. This will be modified
		by the function if IFLG>0
	U : (input/output) Array of length M, which should specify the
		initial approximation to eigenvector. After execution it
		will contain the calculated eigenvector.
	IFLG : (input) Integer variable to specify the type of iteration required
		If IFLG=0 the shift P is kept fixed
		If IFLG=1 the shift P is varied using Rayleigh quotient
		If IFLG=2 the shift P is varied using max(V_s+1)
	EI : (output) Estimated eigenvalue using simple inverse iteration
	ERC : (output) Estimated eigenvalue using Rayleigh quotient
	REPS : (input) Required absolute accuracy. Iteration is terminated
		when all components of eigenvector and the eigenvalue have
		converged to REPS.
	NIT : (input/output) Number of iterations required. If it is
		zero or negative NIT is set to NIT0=100
		
	Error status is returned by the value of the function INVIT.
		0 value implies successful execution
		106 implies that M<=1 or M>IA, in which case no
			calculations are done
		141 implies that vector is zero at some stage and
			calculations are aborted
		142 implies that inverse iteration has failed to converge

	Required functions : GAUELM
*/

#include <math.h>
#include <stdlib.h>

int gauelm(int n, int num, double * a, double * x, double * det,
       int inc[], int lj, int * iflg);

int invit(double *a, int m, int ia, double *p, double u[], int iflg,
	double *ei, double *erc, double reps, int *nit)

{
	int i,j,k,ifl,km,ier, nit0=100;
	double epi,r1,ukm,s1,s2,det;
	int *iwk;
	double *wk,*v;

	if(m<=1 || m> ia) return 106;

	iwk=(int *) calloc((size_t) m, sizeof(int));
	v=(double *) calloc((size_t) m, sizeof(double));
	wk=(double *) calloc((size_t) (m*m), sizeof(double));

/*	Copy the matrix to WK and apply the shift */
	for(i=0; i<m; ++i) {
		v[i]=u[i];
		for(j=0; j<m; ++j) wk[i+j*m]=a[i+j*m];
		wk[i+i*m]=a[i+i*m]-(*p);
	}

	ifl=1;
	if(*nit<=0) *nit=nit0;
/*	Perform Gaussian elimination on A-pI */
	ier=gauelm(m,1,wk,u,&det,iwk,m,&ifl);
	if(ier>0) {free(wk); free(v); free(iwk); return ier;}

	epi=0.0;
/*	Loop for inverse iteration */
	for(j=1; j<=(*nit); ++j) {
		ier=gauelm(m,1,wk,u,&det,iwk,m,&ifl);
		if(ier>0) {free(wk); free(v); free(iwk); return ier;}

/*	Normalising the vector U */
		r1=0.0; km=0;
		for(k=0; k<m; ++k) {
			if(r1<fabs(u[k])) {r1=fabs(u[k]); km=k;}
		}
		ukm=u[km];
/*	If the vector is zero, then quit */
		if(ukm==0.0) {free(wk); free(v); free(iwk); return 141;}

/*	The eigenvalue */
		*ei=v[km]/ukm+(*p);
		s1=0.0; s2=0.0;
/*	Calculating the Rayleigh quotient */
		for(k=0; k<m; ++k) {
			s1=s1+u[k]*v[k];
			s2=s2+fabs(u[k])*fabs(u[k]);
			u[k]=u[k]/ukm;
		}
		*erc=(*p)+s1/s2;

/*	Convergence check */
		r1=fabs(*ei-epi);
		for(i=0; i<m; ++i) {
			if(r1<fabs(v[i]-u[i])) r1=fabs(v[i]-u[i]);
			v[i]=u[i];
		}
		if(fabs(r1)<reps) {free(wk); free(v); free(iwk); return 0;}
		epi=(*ei);

		if(iflg>=1) {
/*	Update the shift */
			*p=(*erc);
			if(iflg==2) *p=(*ei);
/*	Setting up the new matrix A-pI */
			for(i=0; i<m; ++i) {
				for(k=0; k<m; ++k) wk[i+k*m]=a[i+k*m];
				wk[i+i*m]=a[i+i*m]-(*p);
			}
			ifl=1;
			ier=gauelm(m,1,wk,u,&det,iwk,m,&ifl);
			if(ier>0) {free(wk); free(v); free(iwk); return ier;}
		}
	}

	free(wk); free(v); free(iwk);
	return 142;
}



/*	Real eigenvalues and left eigenvector of a real matrix using inverse iteration

	A : (input) Array of length IA*M containing the matrix elements
	M : (input) Order of the matrix
	IA : (input) The second dimension of A as declared in the calling function, IA>=M
	P : (input/output) Initial value of the shift. This will be modified
		by the function if IFLG>0
	U : (input/output) Array of length M, which should specify the
		initial approximation to eigenvector. After execution it
		will contain the calculated left eigenvector.
	IFLG : (input) Integer variable to specify the type of iteration required
		If IFLG=0 the shift P is kept fixed
		If IFLG=1 the shift P is varied using Rayleigh quotient
		If IFLG=2 the shift P is varied using max(V_s+1)
	EI : (output) Estimated eigenvalue using simple inverse iteration
	ERC : (output) Estimated eigenvalue using Rayleigh quotient
	REPS : (input) Required absolute accuracy. Iteration is terminated
		when all components of eigenvector and the eigenvalue have
		converged to REPS.
	NIT : (input/output) Number of iterations required. If it is
		zero or negative NIT is set to NIT0=100
		
	Error status is returned by the value of the function INVIT_L.
		0 value implies successful execution
		106 implies that M<=1 or M>IA, in which case no
			calculations are done
		141 implies that vector is zero at some stage and
			calculations are aborted
		142 implies that inverse iteration has failed to converge

	Required functions : GAUELM
*/

#include <math.h>
#include <stdlib.h>

int gauelm(int n, int num, double * a, double * x, double * det,
       int inc[], int lj, int * iflg);

int invit_l(double *a, int m, int ia, double *p, double u[], int iflg,
	double *ei, double *erc, double reps, int *nit)

{
	int i,j,k,ifl,km,ier, nit0=100;
	double epi,r1,ukm,s1,s2,det;
	int *iwk;
	double *wk,*v;

	if(m<=1 || m> ia) return 106;

	iwk=(int *) calloc((size_t) m, sizeof(int));
	v=(double *) calloc((size_t) m, sizeof(double));
	wk=(double *) calloc((size_t) (m*m), sizeof(double));

/*	Copy the matrix to WK and apply the shift */
	for(i=0; i<m; ++i) {
		v[i]=u[i];
		for(j=0; j<m; ++j) wk[j+i*m]=a[i+j*m];
		wk[i+i*m]=a[i+i*m]-(*p);
	}

	ifl=1;
	if(*nit<=0) *nit=nit0;
/*	Perform Gaussian elimination on A-pI */
	ier=gauelm(m,1,wk,u,&det,iwk,m,&ifl);
	if(ier>0) {free(wk); free(v); free(iwk); return ier;}

	epi=0.0;
/*	Loop for inverse iteration */
	for(j=1; j<=(*nit); ++j) {
		ier=gauelm(m,1,wk,u,&det,iwk,m,&ifl);
		if(ier>0) {free(wk); free(v); free(iwk); return ier;}

/*	Normalising the vector U */
		r1=0.0; km=0;
		for(k=0; k<m; ++k) {
			if(r1<fabs(u[k])) {r1=fabs(u[k]); km=k;}
		}
		ukm=u[km];
/*	If the vector is zero, then quit */
		if(ukm==0.0) {free(wk); free(v); free(iwk); return 141;}

/*	The eigenvalue */
		*ei=v[km]/ukm+(*p);
		s1=0.0; s2=0.0;
/*	Calculating the Rayleigh quotient */
		for(k=0; k<m; ++k) {
			s1=s1+u[k]*v[k];
			s2=s2+fabs(u[k])*fabs(u[k]);
			u[k]=u[k]/ukm;
		}
		*erc=(*p)+s1/s2;

/*	Convergence check */
		r1=fabs(*ei-epi);
		for(i=0; i<m; ++i) {
			if(r1<fabs(v[i]-u[i])) r1=fabs(v[i]-u[i]);
			v[i]=u[i];
		}
		if(fabs(r1)<reps) {free(wk); free(v); free(iwk); return 0;}
		epi=(*ei);

		if(iflg>=1) {
/*	Update the shift */
			*p=(*erc);
			if(iflg==2) *p=(*ei);
/*	Setting up the new matrix A-pI */
			for(i=0; i<m; ++i) {
				for(k=0; k<m; ++k) wk[k+i*m]=a[i+k*m];
				wk[i+i*m]=a[i+i*m]-(*p);
			}
			ifl=1;
			ier=gauelm(m,1,wk,u,&det,iwk,m,&ifl);
			if(ier>0) {free(wk); free(v); free(iwk); return ier;}
		}
	}

	free(wk); free(v); free(iwk);
	return 142;
}

File: antia_numerical_methods/f77/xmp8/invit.f
!     PROGRAM TO FIND ONE EIGENVALUE AND THE CORRESPONDING EIGENVECTOR
!     OF A GENERAL REAL MATRIX USING INVERSE ITERATION
!     IT CALCULATES BOTH LEFT AND RIGHT EIGENVECTORS

      PROGRAM INVERS
      IMPLICIT REAL*8(A-H,O-Z)
!     COMPLEX*16 P,U,W,EI,RC,P0
      DIMENSION A(21,21),U(21),V(21),INT(21),W(21,22),B(21,21)
      DATA ((A(I,J),J=1,4),I=1,4)/-2,2,2,2, -3,3,2,2, -2,0,4,2,-1,0,0,5/
!      DATA ((A(I,J),J=1,4),I=1,4)/4,-5,0,3, 0,4,-3,-5, 5,-3,4,0,3,0,5,4/

51    FORMAT(10X,13HTHE MATRIX IS)
52    FORMAT(10F8.3)
53    FORMAT('   IER =',I4,5X,'IFLG =',I2,5X,'EIGENVALUE =',1PD14.6/
     1       5X,'RAYLEIGH QUOTIENT =',D14.6,5X,'FINAL SHIFT =',D14.6)
54    FORMAT('  EIGENVECTOR =',1P4D14.6/(2X,4D14.6))
55    FORMAT('   INITIAL SHIFT =',1P2D14.6)
56    FORMAT('  LEFT EIGENVECTOR =',1P4D14.6/(2X,4D14.6))

      NUM=100
      IA=21
      REPS=1.D-14

100   PRINT *,'TYPE M=ORDER, P=SHIFT,  IFLG=0/1/2'
      PRINT *,'            (QUITS WHEN M.LE.0)'
      READ *,M,P,IFLG
      IF(M.LE.0) STOP

      WRITE(6,51)
      PRINT *,'TYPE IN THE MATRIX ROW-WISE'
      DO 1000 I=1,M
        U(I)=1.0
        V(I)=1.0
        PRINT *,I,'TH ROW'
        READ *,(A(I,J),J=1,M)
        WRITE(6,52) (A(I,J),J=1,M)
1000  CONTINUE

      PRINT *,'TYPE INITIAL VECTOR'
      READ *,(U(I),I=1,M)
      WRITE(6,55) P
      P0=P
      CALL INVIT(A,M,IA,P,U,IFLG,EI,RC,REPS,W,INT,NUM,IER)
      WRITE(6,53) IER,IFLG,EI,RC,P
      WRITE(6,54) (U(I),I=1,M)

!	Calculate the left eigenvector for the same matrix 
      P=P0
      DO 2000 I=1,M
        U(I)=1.0
        V(I)=1.0
2000  CONTINUE
      CALL INVIT_L(A,M,IA,P,U,IFLG,EI,RC,REPS,W,INT,NUM,IER)
      WRITE(6,53) IER,IFLG,EI,RC,P
      WRITE(6,56) (U(I),I=1,M)
      GO TO 100
      END

!     -----------------------------------------------------

!	Solution of a system of linear equations using Gaussian elimination
!	with partial pivoting
!
!	N : (input) Number of equations to be solved
!	NUM : (input) Number of different sets (each with N equations) of
!	        equations to be solved
!	A : (input/output) The matrix of coefficient of size LJ*N
!	        A(I,J) is the coefficient of x_J in Ith equation
!	     	at output it will contain the triangular decomposition
!	X : (input/output) The matrix containing right hand sides (size LJ*NUM)
!	        X(I,J) is the Ith element of Jth right hand side
!	     	at output it will contain the solutions
!	DET : (output) The determinant of the matrix
!	INC : (output) Integer array of length N containing information about
!		interchanges performed during elimination
!	LJ : (input) First dimension of arrays A and X in calling program
!	IER : (output) Error flag, IER=0 signifies successful execution
!		IER=101 implies (N.LE.0 or N.GT.LJ) 
!		IER=121 implies some pivot turned out to be zero and hence
!			matrix must be nearly singular
!	IFLG : (input) Integer parameter to specify the type of computation required
!		If IFLG.LE.0, both elimination and solution are
!			done and IFLG is set to 2
!		If IFLG=1, only elimination is done and IFLG is set to 2
!		If IFLG.GE.2 only solution is calculated, the triangular
!			decomposition should have been calculated earlier
!
!	Required routines : None

      SUBROUTINE GAUELM(N,NUM,A,X,DET,INC,LJ,IER,IFLG)
      IMPLICIT REAL*8(A-H,O-Z)
!	For complex matrices use the following statements instead
!      IMPLICIT REAL*8(R)
!      IMPLICIT COMPLEX*16(A-H,S-Z)

      DIMENSION A(LJ,N),INC(N),X(LJ,NUM)

      IF(N.LE.0.OR.N.GT.LJ) THEN
        IER=101
        RETURN
      ENDIF

      IER=121
      IF(IFLG.LE.1) THEN
!	Perform elimination

        DET=1.0
        DO 2600 K=1,N-1
!	Find the maximum element in the Kth column
          R1=0.0
          KM=K
          DO 2200 L=K,N
            IF(ABS(A(L,K)).GT.R1) THEN
              R1=ABS(A(L,K))
              KM=L
            ENDIF
2200      CONTINUE

          INC(K)=KM
          IF(KM.NE.K) THEN
!	Interchange the rows if needed
            DO 2300 L=K,N
              T1=A(K,L)
              A(K,L)=A(KM,L)
2300        A(KM,L)=T1
            DET=-DET
          ENDIF

          DET=DET*A(K,K)
          IF(A(K,K).EQ.0.0) RETURN
!	To check for singular or nearly singular matrices replace this
!	statement by, where REPS is approximately \hcross*Max(A(I,J))
!         IF(ABS(A(K,K)).LT.REPS) RETURN
          DO 2500 L=K+1,N
            A(L,K)=A(L,K)/A(K,K)
            DO 2500 L1=K+1,N
2500      A(L,L1)=A(L,L1)-A(L,K)*A(K,L1)
2600    CONTINUE
        DET=DET*A(N,N)
        INC(N)=N
!	If pivot is zero then return, IER has been set to 121
        IF(A(N,N).EQ.0.0) RETURN
!	To check for singular or nearly singular matrices replace this
!	statement by, where REPS is approximately \hcross*Max(A(I,J))
!         IF(ABS(A(N,N)).LT.REPS) RETURN

        IER=0
        IF(IFLG.EQ.1) THEN
          IFLG=2
          RETURN
        ENDIF
        IFLG=2
      ENDIF

      IER=0
!	Solution for the NUM different right-hand sides
      DO 5000 J=1,NUM
        DO 3000 K=1,N-1
!	Forward substitution
          IF(K.NE.INC(K)) THEN
            T1=X(K,J)
            X(K,J)=X(INC(K),J)
            X(INC(K),J)=T1
          ENDIF
          DO 3000 L=K+1,N
3000    X(L,J)=X(L,J)-A(L,K)*X(K,J)

!	back-substitution
        X(N,J)=X(N,J)/A(N,N)
        DO 3300 K=N-1,1,-1
          DO 3200 L=N,K+1,-1
3200      X(K,J)=X(K,J)-X(L,J)*A(K,L)
3300    X(K,J)=X(K,J)/A(K,K)
5000  CONTINUE
      END

!     ------------------------------------------------------

!	Real eigenvalue and eigenvector of a real matrix using inverse iteration
!
!	A : (input) Real array of length IA*M containing the matrix elements
!	M : (input) Order of the matrix
!	IA : (input) The first dimension of A as declared in the calling program
!		IA.GE.M
!	P : (input/output) Initial value of the shift. This will be modified
!		by the program if IFLG>0
!	U : (input/output) Real array of length M, which should specify the
!		initial approximation to eigenvector. After execution it
!		will contain the calculated eigenvector.
!	IFLG : (input) Integer variable to specify the type of iteration required
!		If IFLG=0 the shift P is kept fixed
!		If IFLG=1 the shift P is varied using Rayleigh quotient
!		If IFLG=2 the shift P is varied using max(V_s+1)
!	EI : (output) Estimated eigenvalue using simple inverse iteration
!	ERC : (output) Estimated eigenvalue using Rayleigh quotient
!	REPS : (input) Required absolute accuracy. Iteration is terminated
!		when all components of eigenvector and the eigenvalue have
!		converged to REPS.
!	WK : Real array of length M*(M+1) used as scratch space
!	IWK : Integer array of length M used as scratch space
!	NIT : (input/output) Number of iterations required. If it is
!		zero or negative NIT is set to NIT0=100
!	IER : (output) Error parameter, IER=0 implies successful execution
!		IER=106 implies that M.LE.1 or M>IA, in which case no
!			calculations are done
!		IER=141 implies that vector is zero at some stage and
!			calculations are aborted
!		IER=142 implies that inverse iteration has failed to converge
!
!	Required routines : GAUELM
!
      SUBROUTINE INVIT(A,M,IA,P,U,IFLG,EI,ERC,REPS,WK,IWK,NIT,IER)
      IMPLICIT REAL*8(A-H,O-Z)
      DIMENSION A(IA,M),U(M),WK(M,M+1),IWK(M)
      PARAMETER(NIT0=100)

      IF(M.LE.1.OR.M.GT.IA) THEN
        IER=106
        RETURN
      ENDIF

!	Copy the matrix to WK and apply the shift
      DO 1100 I=1,M
        WK(I,M+1)=U(I)
        DO 1000 J=1,M
1000    WK(J,I)=A(J,I)
1100  WK(I,I)=A(I,I)-P

      NUM=1
      LJ=M
      IFL=1
      IF(NIT.LE.0) NIT=NIT0
!	Perform Gaussian elimination on A-pI
      CALL GAUELM(M,NUM,WK,U,DET,IWK,LJ,IER,IFL)
      IF(IER.GT.0) RETURN

      EPI=0.0
!	Loop for inverse iteration
      DO 5000 J=1,NIT
        CALL GAUELM(M,NUM,WK,U,DET,IWK,LJ,IER,IFL)
        IF(IER.GT.0) RETURN

!	Normalising the vector U
        R1=0.0
        KM=0
        DO 3500 K=1,M
          IF(R1.LT.ABS(U(K))) THEN
            R1=ABS(U(K))
            KM=K
          ENDIF
3500    CONTINUE
        UKM=U(KM)
        IF(UKM.EQ.0.0) THEN
!	If the vector is zero, then quit
          IER=141
          RETURN
        ENDIF

!	The eigenvalue
        EI=WK(KM,M+1)/UKM+P
        S1=0.0
        S2=0.0
!	Calculating the Rayleigh quotient
        DO 4000 K=1,M
          S1=S1+U(K)*WK(K,M+1)
!	For complex eigenvalues use the following statement instead of the
!	preceding one
!         S1=S1+CONJG(U(K))*WK(K,M+1)
          S2=S2+ABS(U(K))**2
          U(K)=U(K)/UKM
4000    CONTINUE
        ERC=P+S1/S2

!	Convergence check
        R1=ABS(EI-EPI)
        DO 4500 I=1,M
          R1=MAX(R1,ABS(WK(I,M+1)-U(I)))
          WK(I,M+1)=U(I)
4500    CONTINUE
        IF(ABS(R1).LT.REPS) RETURN
        EPI=EI

        IF(IFLG.GE.1) THEN
!	Update the shift
          P=ERC
          IF(IFLG.EQ.2) P=EI
!	Setting up the new matrix A-pI
          DO 4700 I=1,M
            DO 4600 K=1,M
4600        WK(K,I)=A(K,I)
4700      WK(I,I)=A(I,I)-P
          IFL=1
          CALL GAUELM(M,NUM,WK,U,DET,IWK,LJ,IER,IFL)
          IF(IER.GT.0) RETURN
        ENDIF

5000  CONTINUE
!	Iteration fails to converge
      IER=142
      END

!	--------------------------------------------------------------

!	Real eigenvalue and left eigenvector of a real matrix using inverse iteration
!
!	A : (input) Real array of length IA*M containing the matrix elements
!	M : (input) Order of the matrix
!	IA : (input) The first dimension of A as declared in the calling program
!	P : (input/output) Initial value of the shift. This will be modified
!		by the program if IFLG>0
!	U : (input/output) Real array of length M, which should specify the
!		initial approximation to eigenvector. After execution it
!		will contain the calculated left eigenvector.
!	IFLG : (input) Integer variable to specify the type of iteration required
!		If IFLG=0 the shift P is kept fixed
!		If IFLG=1 the shift P is varied using Rayleigh quotient
!		If IFLG=2 the shift P is varied using max(V_s+1)
!	EI : (output) Estimated eigenvalue using simple inverse iteration
!	ERC : (output) Estimated eigenvalue using Rayleigh quotient
!	REPS : (input) Required absolute accuracy. Iteration is terminated
!		when all components of eigenvector and the eigenvalue have
!		converged to REPS.
!	WK : Real array of length M*(M+1) used as scratch space
!	IWK : Integer array of length M used as scratch space
!	NIT : (input/output) Number of iterations required. If it is
!		zero or negative NIT is set to NIT0=100
!	IER : (output) Error parameter, IER=0 implies successful execution
!		IER=106 implies that M.LE.1 or M>IA, in which case no
!			calculations are done
!		IER=141 implies that vector is zero at some stage and
!			calculations are aborted
!		IER=142 implies that inverse iteration has failed to converge
!
!	Required routines : GAUELM
!
      SUBROUTINE INVIT_L(A,M,IA,P,U,IFLG,EI,ERC,REPS,WK,IWK,NIT,IER)
      IMPLICIT REAL*8(A-H,O-Z)
      DIMENSION A(IA,M),U(M),WK(M,M+1),IWK(M)
      PARAMETER(NIT0=100)

      IF(M.LE.1.OR.M.GT.IA) THEN
        IER=106
        RETURN
      ENDIF

!	Copy the transposed matrix to WK and apply the shift
      DO 1100 I=1,M
        WK(I,M+1)=U(I)
        DO 1000 J=1,M
1000    WK(J,I)=A(I,J)
1100  WK(I,I)=A(I,I)-P

      NUM=1
      LJ=M
      IFL=1
      IF(NIT.LE.0) NIT=NIT0
!	Perform Gaussian elimination on A-pI
      CALL GAUELM(M,NUM,WK,U,DET,IWK,LJ,IER,IFL)
      IF(IER.GT.0) RETURN

      EPI=0.0
!	Loop for inverse iteration
      DO 5000 J=1,NIT
        CALL GAUELM(M,NUM,WK,U,DET,IWK,LJ,IER,IFL)
        IF(IER.GT.0) RETURN

!	Normalising the vector U
        R1=0.0
        KM=0
        DO 3500 K=1,M
          IF(R1.LT.ABS(U(K))) THEN
            R1=ABS(U(K))
            KM=K
          ENDIF
3500    CONTINUE
        UKM=U(KM)
        IF(UKM.EQ.0.0) THEN
!	If the vector is zero, then quit
          IER=141
          RETURN
        ENDIF

!	The eigenvalue
        EI=WK(KM,M+1)/UKM+P
        S1=0.0
        S2=0.0
!	Calculating the Rayleigh quotient
        DO 4000 K=1,M
          S1=S1+U(K)*WK(K,M+1)
!	For complex eigenvalues use the following statement instead of the
!	preceding one
!         S1=S1+CONJG(U(K))*WK(K,M+1)
          S2=S2+ABS(U(K))**2
          U(K)=U(K)/UKM
4000    CONTINUE
        ERC=P+S1/S2

!	Convergence check
        R1=ABS(EI-EPI)
        DO 4500 I=1,M
          R1=MAX(R1,ABS(WK(I,M+1)-U(I)))
          WK(I,M+1)=U(I)
4500    CONTINUE
        IF(ABS(R1).LT.REPS) RETURN
        EPI=EI

        IF(IFLG.GE.1) THEN
!	Update the shift
          P=ERC
          IF(IFLG.EQ.2) P=EI
!	Setting up the new matrix A-pI
          DO 4700 I=1,M
            DO 4600 K=1,M
4600        WK(K,I)=A(I,K)
4700      WK(I,I)=A(I,I)-P
          IFL=1
          CALL GAUELM(M,NUM,WK,U,DET,IWK,LJ,IER,IFL)
          IF(IER.GT.0) RETURN
        ENDIF

5000  CONTINUE
!	Iteration fails to converge
      IER=142
      END
